File: backend/test/screenplay/actors/actor.py
class Actor:
    def __init__(self, name: str):
        self.name = name
        self._abilities = {}
        self._memory = {}

    def can(self, ability):
        self._abilities[type(ability)] = ability
        return self

    def ability_to(self, ability_type):
        return self._abilities[ability_type]

    def remember(self, key: str, value):
        self._memory[key] = value

    def recall(self, key: str, default=None):
        return self._memory.get(key, default)

    def attempts_to(self, *tasks):
        for task in tasks:
            task.perform_as(self)

File: frontend/test/screenplay/screenplay/actor.ts
import type { Task } from "./task";
import type { Question } from "./question";

export type Ability = {
  readonly page?: unknown;
};

type AbilityCtor<T extends Ability> = new (...args: any[]) => T;

export class Actor {
  private abilities = new Map<Function, Ability>();

  private constructor(private readonly actorName: string) {}

  static named(name: string) {
    return new Actor(name);
  }

  whoCan(...abilities: Ability[]) {
    abilities.forEach((ability) => {
      this.abilities.set(ability.constructor, ability);
    });
    return this;
  }

  abilityTo<T extends Ability>(abilityType: AbilityCtor<T>): T {
    const ability = this.abilities.get(abilityType);
    if (!ability) {
      throw new Error(
        `Actor ${this.actorName} lacks ability ${abilityType.name}`,
      );
    }
    return ability as T;
  }

  async attemptsTo(...tasks: Task[]) {
    for (const task of tasks) {
      await task.performAs(this);
    }
  }

  async asks<T>(question: Question<T>): Promise<T> {
    return question.answeredBy(this);
  }
}
